src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCodeLensProvider.ts

```diff
@@ -0,0 +1,130 @@
+        try {
+            const razorDocument = await this.documentManager.getDocument(document.uri);
+            if (!razorDocument) {
+                return;
+            }
+
+            const csharpDocument = razorDocument.csharpDocument;
+
+            // Get all the code lenses that applies to our projected C# document.
+            const codeLenses = await vscode.commands.executeCommand<vscode.CodeLens[]>(
+                'vscode.executeCodeLensProvider',
+                csharpDocument.uri) as vscode.CodeLens[];
+
+            // Re-map the CodeLens locations to the original Razor document.
+            const remappedCodeLenses = new Array<vscode.CodeLens>();
+            for (const codeLens of codeLenses) {
+                const result = await this.serviceClient.mapToDocumentRange(
+                    LanguageKind.CSharp,
+                    codeLens.range,
+                    razorDocument.uri);
+                if (result) {
+                    const newCodeLens = new RazorCodeLens(result.range, razorDocument.uri, document, codeLens.command);
+                    remappedCodeLenses.push(newCodeLens);
+                } else {
+                    // This means this CodeLens was for non-user code. We can safely ignore those.
+                }
+            }
+
+            return remappedCodeLenses;
+
+        } catch (error) {
+            return [];
+        }
+    }
+
+    public async resolveCodeLens(codeLens: vscode.CodeLens, token: vscode.CancellationToken) {
+        if (codeLens instanceof RazorCodeLens) {
+            return await this.resolveRazorCodeLens(codeLens, token);
+        }
+    }
+
+    private async resolveRazorCodeLens(codeLens: RazorCodeLens, token: CancellationToken): Promise<vscode.CodeLens> {
+        // Initialize with default values.
+        codeLens.command = {
+            title: '',
+            command: '',
+            arguments: []
+        };
+
+        try {
+            const razorDocument = await this.documentManager.getDocument(codeLens.uri);
+            if (!razorDocument) {
+                return codeLens;
+            }
+
+            // Make sure this CodeLens is for a valid location in the projected C# document.
+            const projection = await this.getProjection(codeLens.document, codeLens.range.start, token);
+            if (!projection || projection.languageKind != LanguageKind.CSharp) {
+                return codeLens;
+            }
+
+            const references = await vscode.commands.executeCommand<vscode.Location[]>(
+                'vscode.executeReferenceProvider',
+                projection.uri,
+                projection.position) as vscode.Location[];
+
+            // We now have a list of references to show in the CodeLens.
+            // Let's remap any locations pointing to any background documents back to the original documents.
+            // Technically, Omnisharp should have already remapped those. We're just being extra cautious here.
```

Comment:
👍

Reply:
Yeah I like the descriptive comments since some of these `executeCommand`-based flows can be a bit hard to follow.